Google signing for React Native apps (#18)
bcgov/mobile-developer-guide: docs/google_react_native_signing.md

# Google signing for React Native apps

> Refer to React Native's [Publishing to Google Play Store](https://reactnative.dev/docs/signed-apk-android) guide for an overview of app signing. That guide describes how to sign an app from a developer's computer. This page describes how to setup those steps in a CI/CD pipeline using GitHub Actions.

Reviewer: ```suggestion
Learn how to setup Google signing in a CI/CD pipeline using GitHub Actions.

Need a general overview of what app signing is? Review React Native's [Publishing to Google Play Store](https://reactnative.dev/docs/signed-apk-android) guide which describes how to sign an app from a developer's computer. 
```

Reviewer: Always start the page with an intro sentence on what the content is about

> ## Prerequisite

Reviewer: It would be nice to see this page restructured to use steps/numbers to walk someone through what they need to do.

Author: I added the steps at the top of the page, with links to the sections. I wasn't sure how to restructure it into steps, as each step had a fair bit of detail.

Your app must be setup to use [Google Play App Signing](https://developer.android.com/studio/publish/app-signing#app-signing-google-play).

## Create the upload key
Follow React Native's [instructions to generate an upload key](https://reactnative.dev/docs/signed-apk-android#generating-an-upload-key).

## Setup GitHub secrets

> In your app's GitHub repo, setup the following secrets:

Reviewer: ```suggestion
In your app's GitHub repo, setup 4 secrets:
```

* ANDROID_UPLOAD_KEYSTORE_BASE64={file-name-from-above-step}.keystore - encoded as base64
* ANDROID_KEY_ALIAS={alias-from-above-step}
* ANDROID_SIGNING_STORE_PASSWORD={password-from-above-step}
* ANDROID_SIGNING_KEY_PASSWORD={password-from-above-step}

> Use the following command to encode the keystore file as base64:

Reviewer: ```suggestion
To encode the keystore file as base64, use this command:
```

`base64 -i your-keystore-file.keystore | pbcopy`

## Update Gradle Files 

In the `android/app/build.gradle` file add a `release` config under `signingConfigs`. 

```js
signingConfigs {
        debug {
            storeFile file('debug.keystore')
            storePassword 'android'
            keyAlias 'androiddebugkey'
            keyPassword 'android'
        }
        release {
            // This will look for the release.keystore file in the android/app/ folder. The release.keystore file will be created in the GitHub Action
            storeFile = file("release.keystore")
            // The environment keys must match the secret keys setup in GitHub
            keyAlias System.getenv("ANDROID_KEY_ALIAS")
            storePassword System.getenv("ANDROID_SIGNING_STORE_PASSWORD")
            keyPassword System.getenv("ANDROID_SIGNING_KEY_PASSWORD")
        }
    }
```

> Additionally, we need to tell Gradle to use our new config for release builds:

Reviewer: ```suggestion
You also need to tell Gradle to use the new config for release builds:
```

```js
    buildTypes {
        debug {
            signingConfig signingConfigs.debug
        }
        release {
            // Using release from above
            signingConfig signingConfigs.release
        }
    }    
```

Optionally, set the versionCode automatically from GitHub Actions 
```js
defaultConfig {
        ...
        versionCode System.getenv("VERSION_CODE") ? System.getenv("VERSION_CODE").toInteger() : 1
        versionName "1.1"
        ...
    }
```

## GitHub Action

>  See the [bc-mobile-wallet](https://github.com/bcgov/bc-wallet-mobile) project for a complete GitHub Action workflow. Below is a basic GitHub Action.

Reviewer: ```suggestion
Review the [bc-mobile-wallet](https://github.com/bcgov/bc-wallet-mobile) project for a complete GitHub Action workflow. 

Example of a basic GitHub Action:
```

      - name: Decode keystore file
        env:
          UPLOAD_KEYSTORE: ${{ secrets.ANDROID_UPLOAD_KEYSTORE_BASE64 }}
        run: |
          echo "${UPLOAD_KEYSTORE}" | base64 -d > ${KEYSTORE_FILE}

      - name: Sign app 
        env:
          # These environment keys MUST match the keys in the build.gradle file
          ANDROID_KEY_ALIAS: ${{ secrets.ANDROID_KEY_ALIAS }}
          ANDROID_SIGNING_KEY_PASSWORD: ${{ secrets.ANDROID_SIGNING_KEY_PASSWORD }}
          ANDROID_SIGNING_STORE_PASSWORD: ${{ secrets.ANDROID_SIGNING_STORE_PASSWORD }}
          # Automatically update the VersionCode number in the build.gradle file. The number is not committed to git.
          VERSION_CODE: ${{ github.run_number }}
        run: |
            npx react-native build-android --mode=release
       
      - name: Upload Artifact
        uses: actions/upload-artifact@v3
        with:
            name: Signed App Bundle
            # This is the default path where the bundle is created
            path: android/app/build/outputs/bundle/release/app-release.aab
            if-no-files-found: error
            retention-days: 7    
```

## Upload the build

### GitHub Action Upload

> See the [Publish Your App](publish.md) page for information on how to upload from your CI/CD pipeline.

Reviewer: ```suggestion
Review the [Publish Your App](publish.md) page for information on how to upload from your CI/CD pipeline.
```

Reviewer: again, avoid use of "see"
